### PR TITLE
[LAA] Check accesses don't overlap early to determine NoDep

### PR DESCRIPTION
**Pull request by fhahn**
Use getStartAndEndForAccess to compute the start and end of both src 
and sink (factored out to helper in bce3680f45b57f). If they do not
overlap (i.e. SrcEnd <= SinkStart || SinkEnd <= SrcStart), there is no
dependence, regardless of stride.

### CODE REVIEW

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 1]
```diff
@@ -1991,13 +2015,12 @@ getDependenceDistanceStrideAndSize(
     return MemoryDepChecker::Dependence::Unknown;
   }
 
-  uint64_t TypeByteSize = DL.getTypeAllocSize(ATy);
   bool HasSameSize =
       DL.getTypeStoreSizeInBits(ATy) == DL.getTypeStoreSizeInBits(BTy);
```
Comment (Meinersbur) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2019]: There is a difference now with the handling of scalable types. `getTypeStoreSizeInBits`/`getTypeAllocSize` return the minimum size. Now `TypeByteSize` is 0 if `ATy` is a scalable type (and happens to have the same minimum size as the `TypeStoreSize` of `BTy`), but not if the other way around with `BTy`.

I think it's safer to pass `0` for such cases (minimum size is probably not safe for a scalable type, why wasn't the maximum size used?), so maybe do the same for `BTy`?

**Review by Meinersbur — APPROVED** [review]
LGTM

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp:2019]: Thanks, I've updated the code to only query `DL.getTypeAllocSize` if non of the accessed types is scalable.

At the moment, those sizes will only be used by the newly added code; `getPtrStride` will return 0 for scalable types, so we will for now always return an unknown dependence due to the stride checks above.

**Comment by preames** [comment]
LGTM.  Thanks for generalizing.  The result is much cleaner and easier to follow.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 2]
```diff
@@ -213,16 +213,14 @@ getStartAndEndForAccess(const Loop *Lp, const SCEV *PtrExpr, Type *AccessTy,
...
-    assert(AR && "Invalid addrec expression");
+  } else if (auto *AR = dyn_cast<SCEVAddRecExpr>(PtrExpr)) {
     const SCEV *Ex = PSE.getBackedgeTakenCount();
 
     ScStart = AR->getStart();
     ScEnd = AR->evaluateAtIteration(Ex, *SE);
     const SCEV *Step = AR->getStepRecurrence(*SE);
 
-    // For expressions with negative step, the upper bound is ScStart and the
+    // For expressions with negative step, the& upper bound is ScStart and the
```
Comment (Meinersbur) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:223]: [nit]

**Review by Meinersbur — APPROVED** [review]
LGTM

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp]: dropped, thanks!

**Comment by nikic** [comment]
It looks like this causes large compile-time regressions in some cases: http://llvm-compile-time-tracker.com/compare.php?from=7cee61c82f7fe5a563c2781e855c13238c44931f&to=1b377dbeb73ef3abec246fe11fc98ec699625c0c&stat=instructions%3Au

For example, libclamav_phishcheck.c regresses by more than 20%.

**Comment by fhahn** [comment]
oh wow that's definitely unexpected. Let me take a look

**Comment by nikic** [comment]
@fhahn Any news on this? I'd suggest reverting the change for now -- with the amount of commits in between it will be hard to determine whether the regression was fully addressed.

**Comment by fhahn** [comment]
> @fhahn Any news on this? I'd suggest reverting the change for now -- with the amount of commits in between it will be hard to determine whether the regression was fully addressed.

compile-time should be restored by limiting to loop-invariant source/sinks for now (234cc40adc610a55d1a5a2fe798a9dd07b993f0c). There's only a very small change with the original commit reverted on top of that one (0.01% change). WDYT?

 I also have a few patches to improve compile-time for the general version which I'll be working on as follow-ups

**Comment by nikic** [comment]
That looks good to me, thanks!

**Pull request merged**